File: harness-engineering/src/lib/chat/permission-queue.ts
const pendingPermissions = new Map<string, {
  engine: { resolvePermission: (optionId: string) => void }
}>()

export function registerPendingPermission(
  requestId: string,
  engine: { resolvePermission: (optionId: string) => void },
): void {
  pendingPermissions.set(requestId, { engine })
}

export function resolvePermission(requestId: string, optionId: string): boolean {
  const entry = pendingPermissions.get(requestId)
  if (!entry) return false
  entry.engine.resolvePermission(optionId)
  pendingPermissions.delete(requestId)
  return true
}

File: harness-engineering-py/backend/app/services/permission_queue.py
from typing import Any, Dict


_pending_permissions: Dict[str, Any] = {}


def register_pending_permission(request_id: str, engine: Any) -> None:
    _pending_permissions[request_id] = engine


def resolve_permission(request_id: str, option_id: str) -> bool:
    engine = _pending_permissions.pop(request_id, None)
    if engine is None:
        return False
    if hasattr(engine, "resolve_permission"):
        engine.resolve_permission(option_id)
    return True
